prog/ai/inference_endpoint_replica_nexus.rb

```diff
@@ -235,4 +253,113 @@ def update_billing_records(project_usage)
+    case inference_endpoint.engine
+    when "vllm"
+      env = (inference_endpoint.gpu_count == 0) ? "vllm-cpu" : "vllm"
+      "/opt/miniconda/envs/#{env}/bin/vllm serve /ie/models/model --served-model-name #{inference_endpoint.model_name} --disable-log-requests --host 127.0.0.1 #{inference_endpoint.engine_params}"
+    when "runpod"
+      "ssh -N -L 8000:localhost:8000 root@#{inference_endpoint_replica.external_state["ip"]} -p #{inference_endpoint_replica.external_state["port"]} -i /ie/workdir/.ssh/runpod -o UserKnownHostsFile=/ie/workdir/.ssh/known_hosts -o StrictHostKeyChecking=accept-new"
+    else
+      fail "BUG: unsupported inference engine"
+    end
+  end
+
+  def create_runpod_pod
+    response = Excon.post("https://api.runpod.io/graphql",
+      headers: {"content-type" => "application/json", "authorization" => "Bearer #{Config.runpod_api_key}"},
+      body: {"query" => "query Pods { myself { pods { id name runtime  { ports { ip isIpPublic privatePort publicPort type } } } } }"}.to_json,
+      expects: 200)
+
+    pods = JSON.parse(response.body)["data"]["myself"]["pods"]
+    pod = pods.find { |pod| pod["name"] == inference_endpoint_replica.ubid }
+
+    return pod["id"] if pod
+
+    ssh_keys = vm.sshable.cmd(<<-CMD
+if ! sudo test -f /ie/workdir/.ssh/runpod; then  
+  sudo -u ie mkdir -p /ie/workdir/.ssh
+  sudo -u ie ssh-keygen -t ed25519 -C #{inference_endpoint_replica.ubid}@ubicloud.com -f /ie/workdir/.ssh/runpod -N '' -q
+fi
+sudo cat /ie/workdir/.ssh/runpod.pub
+    CMD
+                             ) + Config.operator_ssh_public_keys
+
+    vllm_params = "--served-model-name #{inference_endpoint.model_name} --disable-log-requests --host 127.0.0.1 #{inference_endpoint.engine_params}"
+
+    graphql_query = <<~GRAPHQL
+      mutation {
+        podFindAndDeployOnDemand(
+          input: {
+            cloudType: ALL
+            dataCenterId: "#{inference_endpoint.external_config["data_center"]}"
+            gpuCount: #{inference_endpoint.external_config["gpu_count"]}
+            gpuTypeId: "#{inference_endpoint.external_config["gpu_type"]}"
+            containerDiskInGb: #{inference_endpoint.external_config["disk_gib"]}
+            minVcpuCount: #{inference_endpoint.external_config["min_vcpu_count"]}
+            minMemoryInGb: #{inference_endpoint.external_config["min_memory_gib"]}
+            imageName: "#{inference_endpoint.external_config["image_name"]}"
+            name: "#{inference_endpoint_replica.ubid}"
+            volumeInGb: 0
+            ports: "22/tcp"
+            env: [
+              { key: "HF_TOKEN", value: "#{Config.huggingface_token}" },
+              { key: "HF_HUB_ENABLE_HF_TRANSFER", value: "1"},
+              { key: "MODEL_PATH", value: "/model"},
+              { key: "MODEL_NAME_HF", value: "#{inference_endpoint.external_config["model_name_hf"]}"},
+              { key: "VLLM_PARAMS", value: "#{vllm_params}"},
+              { key: "SSH_KEYS", value: "#{ssh_keys.gsub("\n", "\\n")}" }
```

Comment:
nit: https://github.com/ubicloud/ubicloud/pull/2977#discussion_r1993661643

```suggestion

    config = inference_endpoint.external_config
    graphql_query = <<~GRAPHQL
      mutation {
        podFindAndDeployOnDemand(
          input: {
            cloudType: ALL
            dataCenterId: "#{config["data_center"]}"
            gpuCount: #{config["gpu_count"]}
            gpuTypeId: "#{config["gpu_type"]}"
            containerDiskInGb: #{config["disk_gib"]}
            minVcpuCount: #{config["min_vcpu_count"]}
            minMemoryInGb: #{config["min_memory_gib"]}
            imageName: "#{config["image_name"]}"
            name: "#{inference_endpoint_replica.ubid}"
            volumeInGb: 0
            ports: "22/tcp"
            env: [
              { key: "HF_TOKEN", value: "#{Config.huggingface_token}" },
              { key: "HF_HUB_ENABLE_HF_TRANSFER", value: "1"},
              { key: "MODEL_PATH", value: "/model"},
              { key: "MODEL_NAME_HF", value: "#{config["model_name_hf"]}"},
              { key: "VLLM_PARAMS", value: "#{vllm_params}"},
              { key: "SSH_KEYS", value: "#{ssh_keys.gsub("\n", "\\n")}" }
```